internal/runners/packages/searchView.go

```diff
@@ -0,0 +1,304 @@
+package packages
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/charmbracelet/bubbles/viewport"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+	"golang.org/x/term"
+)
+
+var (
+	keyName        = locale.Tl("search_name", "Name")
+	keyDescription = locale.Tl("search_description", "Description")
+	keyWebsite     = locale.Tl("search_website", "Website")
+	keyLicense     = locale.Tl("search_license", "License")
+	keyVersions    = locale.Tl("search_versions", "Versions")
+	keyVulns       = locale.Tl("search_vulnerabilities", "Vulnerabilities (CVEs)")
+
+	keys = []string{
+		keyName,
+		keyDescription,
+		keyWebsite,
+		keyLicense,
+		keyVersions,
+		keyVulns,
+	}
+)
+
+const (
+	leftPad        = 2
+	verticalMargin = 7
+	scrollUp       = "up"
+	scrollDown     = "down"
+)
+
+type errMsg error
+
+type view struct {
+	width         int
+	height        int
+	remaining     int
+	content       string
+	index         map[string]int
+	searchResults *structuredSearchResults
+	ready         bool
+	err           error
+	viewport      viewport.Model
+}
+
+func NewView(results *structuredSearchResults, out output.Outputer) (*view, error) {
+	outFD, ok := out.Config().OutWriterFD()
+	if !ok {
+		logging.Error("Could not get output writer file descriptor, falling back to stdout")
+		outFD = os.Stdout.Fd()
+	}
+
+	width, height, err := term.GetSize(int(outFD))
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not get terminal size")
+	}
+
+	return &view{
+		width:         width,
+		height:        height,
+		searchResults: results,
+	}, nil
+}
+
+func (v *view) Init() tea.Cmd {
+	return nil
+}
+
+func (v *view) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var (
+		cmd  tea.Cmd
+		cmds []tea.Cmd
+	)
+
+	switch msg := msg.(type) {
```

Comment:
It looks like scrolling is already handled through the viewport lib:

https://github.com/charmbracelet/bubbles/blob/master/viewport/viewport.go#L287

Is there a reason we can't make use of that?

Reply:
Took some time to try and use this. The `Update` method does not implement the correct interface so it's not a simple replacement. We would likely have to do some wrapping. I didn't realize the viewport had these methods as I was following the example that I linked in my initial comment.

Reply:
Can't you just call the viewport's `Update()` method from your `Update()` method? ie. handle whatever you need to handle, then defer the rest to the viewport package.

Reply:
Took another look at this and no, it's not that simple. Again, the `Update()` function return a struct that doesn't satisfy the `tea.Model` interface. This means that we can't just return the values that the `Update()` function returns.

I didn't spend _a lot_ of time trying to get this working but I managed to by replacing the underlying viewport. Unfortunately, this means we lose the header and footer spacing that we want to setup so it looks a bit off.

Looking at bit more at the code you linked it's not immediately evident what benefit we get from using this `Update` function. It would be handling the scrolling keys for us but that's about it.

If there is some other benefit or if this is something that we need before this can be merged I can try to get it working but it will take some time.